components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/dao/impl/AuthenticatorManagementDAOImpl.java

```diff
@@ -0,0 +1,340 @@
+    public static final String LOCAL_IDP_NAME = "LOCAL";
+
+    public AuthenticatorManagementDAOImpl() {
+    }
+
+    @Override
+    public LocalAuthenticatorConfig addUserDefinedLocalAuthenticator(LocalAuthenticatorConfig authenticatorConfig,
+                Integer tenantId, AuthenticationType type) throws AuthenticatorMgtException {
+
+        Connection dbConnection = IdentityDatabaseUtil.getDBConnection(true);
+
+        try (NamedPreparedStatement statement = new NamedPreparedStatement(dbConnection, Query.ADD_AUTHENTICATOR_SQL)) {
+
+            statement.setString(Column.NAME, authenticatorConfig.getName());
+            statement.setString(Column.DISPLAY_NAME, authenticatorConfig.getDisplayName());
+            statement.setString(Column.DEFINED_BY, authenticatorConfig.getDefinedByType().toString());
+            statement.setString(Column.AUTHENTICATION_TYPE, type.toString());
+            int isEnabled = authenticatorConfig.isEnabled() ? 1 : 0;
+            statement.setInt(Column.IS_ENABLED, isEnabled);
+            statement.setString(Column.IDP_NAME, LOCAL_IDP_NAME);
+            statement.setInt(Column.TENANT_ID, tenantId);
+            statement.executeUpdate();
+
+            if (authenticatorConfig.getProperties() != null) {
+
+                int configId = getAuthenticatorIdentifier(dbConnection, authenticatorConfig.getName(), tenantId);
+                try (NamedPreparedStatement statementProp = new NamedPreparedStatement(dbConnection,
+                        Query.ADD_AUTHENTICATOR_PROP_SQL)) {
+
+                    for (Property property : authenticatorConfig.getProperties()) {
+                        statementProp.setInt(Column.AUTHENTICATOR_ID, configId);
+                        statementProp.setInt(Column.TENANT_ID, tenantId);
+                        statementProp.setString(Column.PROPERTY_KEY, property.getName());
+                        statementProp.setString(Column.PROPERTY_VALUE, property.getValue());
+                        if (property.isConfidential()) {
+                            statementProp.setString(Column.IS_SECRET, IS_TRUE_VALUE);
+                        } else {
+                            statementProp.setString(Column.IS_SECRET, IS_FALSE_VALUE);
+                        }
+                        statementProp.executeUpdate();
+                    }
+                }
+            }
+            IdentityDatabaseUtil.commitTransaction(dbConnection);
+
+            return getUserDefinedLocalAuthenticatorByName(dbConnection, authenticatorConfig.getName(), tenantId);
+        } catch (SQLException | AuthenticatorMgtException e) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(String.format("Error while adding the Authenticator of named: %s in Tenant Domain: %s. " +
+                                "Rolling back added Authenticator information.", authenticatorConfig.getName(),
+                                IdentityTenantUtil.getTenantDomain(tenantId)));
```

Comment:
Error while adding the authenticator: %s in tenant domain: %s.

Reply:
Addressed with commit d72ebd31b7150e7c562faf83ee80300a60e1e5d8